### ISSUE TITLE
twisted.conch.test.test_transport.SSHCiphersTestCase.test_getCipher() fails with pycrypto 2.6

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5700]
|<img alt="Arfrever's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Arfrever reported|
|-|-|
|Trac ID|trac#5700|
|Type|defect|
|Created|2012-06-06 02:13:29Z|
|Branch|https://github.com/twisted/twisted/tree/getCipher-pycrypto26-5700|

twisted.conch.test.test_transport.SSHCiphersTestCase.test_getCipher() fails with pycrypto 2.6 and passes with pycrypto 2.5. I use Twisted 12.1.0.

```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/var/tmp/portage/dev-python/twisted-conch-12.1.0-r1000/temp/usr/lib64/python2.7/site-packages/twisted/conch/test/test_transport.py", line 1832, in test_getCipher
    self.assertTrue(str(cip).startswith('<' + modName))
twisted.trial.unittest.FailTest: None

twisted.conch.test.test_transport.SSHCiphersTestCase.test_getCipher
-------------------------------------------------------------------------------
```

If I add print((cipName, cip, modName)) in the loop in this test, then I see:
```
('cast128-ctr', <Crypto.Cipher.CAST.CAST128Cipher instance at 0x2f50680>, 'CAST')
```

Attachments:

* [twisted-conch-test-getCipher.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/408/408805b4e4a2e02652490c1b606c928208a4d54f/96666fe3029002382cba09f7ed1f6959159ed278.patch) (914 bytes) - added by marienz on 2012-06-08 14:16:07Z - make test_getCiphers work with pycrypto 2.5 and 2.6

<details><summary>Searchable metadata</summary>

```
trac-id__5700 5700
type__defect defect
reporter__Arfrever Arfrever
priority__normal normal
milestone__ 
branch__branches_getCipher_pycrypto26_5700 branches/getCipher-pycrypto26-5700
branch_author__z3p z3p
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1338948809000000 1338948809000000
changetime__1341771286000000 1341771286000000
version__None None
owner__z3p z3p
cc__z3p cc__Arfrever cc__marienz
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

The test is rather dump, but the error comes from a name change in pycrypto classes, from "<CAST object at xxx>" to "<Crypto.Cipher.CAST.CAST128Cipher instance at xxx>". It should be easy to fix, if we can a buildslave with the updated pycrypto version.

**Comment by twisted-trac** [comment]
|<img alt="marienz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|marienz commented|
|-|-|

The attached patch works for me against both pycrypto 2.5 and pycrypto 2.6. Alternatively the test could just check if modName is in repr(cip), but the attached change seems as safe and makes the test a little more precise.

The actual change to pycrypto is that 2.5 gave you an instance of an extension type while 2.6 gives you an instance of an old-style class. So if you want to work with both and don't want to do a fairly fuzzy match against repr(cip) you really need getClass here (it's type(cip) on one and cip.__class__ on the other).

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|z3p commented|
|-|-|

(In [34677]) Branching to 'getCipher-pycrypto26-5700'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|z3p commented|
|-|-|

Applied the patch, and forced a build: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/getCipher-pycrypto26-5700

Normally I'd review it here, but I have a headache and don't trust myself. If no one gets to it, I'll check it again later.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to z3p|
|-|-|

Looks good to me, please merge. We still need a build slave with the new version of the library, though. Can you send an email to the list or ping the buildbot maintainer to make sure it happens? Thanks.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|z3p set status to `closed`|
|-|-|

(In [34697]) Merge getCipher-pycrypto26-5700: fix test failure with PyCrypto 2.6

Author: marienz
Reviewer: therve
Fixes: [#5700](https://github.com/twisted/twisted/issues/5700)

PyCrypto 2.6 changed the str() of some of its classes.  This caused a failure
in a test which was checking the str() to verify that we received the correct
encryption class.  This branch changes the test to check the __name__ of the
received class.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|z3p commented|
|-|-|

I've also sent an e-mail to the Twisted list to inquire about a new buildbox.
